shared/src/hooks/useWeb3Wallet.tsx

```diff
@@ -0,0 +1,173 @@
+  connectedWallet: Wallet | undefined;
+  ethereumProvider: providers.Web3Provider | undefined;
+  solanaWallet: SolanaWalletInterface | undefined;
+}
+
+export const useWeb3Wallet = (): Web3WalletData => {
+  const [chain, setChain] = useChain();
+  const [chainToSwitch, setChainToSwitch] = useState<Chains | null>(null);
+  const [connectingWallet, setConnectingWallet] = useState<Wallet>();
+  const [connectedWallet, setConnectedWallet] = useState<Wallet>();
+
+  const {
+    chainId: chainIdEth,
+    active: isActiveEth,
+    activate: activateEth,
+    account: accountEth,
+    library: libraryEth,
+    deactivate: _deactivateEth,
+    connector: connectorEth,
+  } = useEVMWallet();
+
+  const {
+    wallet: walletSolana,
+    connected: isActiveSolana,
+    disconnect: deactivateSolana,
+    publicKey: publicKeySolana,
+    select: selectWalletSolana,
+  } = useSolanaWallet();
+
+  // This hook manages the entire multi-chain flow
+  // It checks when an active provider is available and sets the global chainContext accordingly
+  useEffect(() => {
+    if (isActiveEth && chainIdEth) {
+      setChain(ID_TO_CHAINS[chainIdEth as number]);
+    } else if (isActiveSolana && !chainIdEth) {
+      setChain(Chains.Solana);
+    }
+  }, [chainIdEth, isActiveEth, isActiveSolana]);
+
+  // This hook checks if there is an EVM chainId to switch to
+  // If so, it will prompt switchChains only when a provider is available
+  useEffect(() => {
+    if (chainToSwitch && libraryEth) {
+      switchChains(libraryEth, CHAINS_TO_ID[chainToSwitch]).then(() => {
+        setChainToSwitch(null);
+      });
+    }
+  }, [libraryEth, chainToSwitch]);
+
+  const deactivateEth = useCallback(async () => {
+    if (connectorEth) {
+      if (connectorEth instanceof WalletConnectConnector) {
+        await connectorEth.close();
+      }
+
+      _deactivateEth();
+      setConnectedWallet(undefined);
+    }
+  }, [connectorEth, _deactivateEth]);
+
+  const activate = useCallback(
+    async (wallet: Wallet, chain: Chains) => {
+      setConnectingWallet(wallet);
+
+      try {
+        if (isEthereumWallet(wallet)) {
+          if (!isActiveEth) {
+            const connector = evmConnectors[wallet as EthereumWallet]();
+            await activateEth(connector);
+          }
+
+          await deactivateSolana().then(() => {
+            setChainToSwitch(chain);
+            setConnectedWallet(wallet as EthereumWallet);
+          });
+        } else if (isSolanaWallet(wallet)) {
+          let walletName: WalletName = WalletName.Phantom;
+          switch (wallet) {
+            case SolanaWallet.Phantom:
+              walletName = WalletName.Phantom;
+              break;
+            case SolanaWallet.Solflare:
+              walletName = WalletName.Solflare;
+              break;
+          }
+
+          await deactivateEth().then(() => {
+            selectWalletSolana(walletName);
+            setConnectedWallet(wallet as SolanaWallet);
+          });
```

Comment:
We don't have to do a `then()` right? Because `await` implicitly enforces the order

Reply:
Ah did not notice this part here, thanks for pointing out